presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionTask.java

```diff
@@ -28,71 +24,55 @@
-import com.facebook.presto.server.SimpleHttpResponseHandler;
-import com.facebook.presto.server.TaskUpdateRequest;
-import com.facebook.presto.server.remotetask.RemoteTaskStats;
 import com.facebook.presto.server.smile.BaseResponse;
+import com.facebook.presto.spark.execution.http.PrestoSparkHttpWorkerClient;
+import com.facebook.presto.spi.page.SerializedPage;
 import com.facebook.presto.sql.planner.PlanFragment;
 import com.google.common.util.concurrent.FutureCallback;
 import com.google.common.util.concurrent.Futures;
-import com.google.common.util.concurrent.ListenableFuture;
-import org.joda.time.DateTime;
-
-import javax.annotation.concurrent.GuardedBy;
+import org.checkerframework.checker.nullness.qual.Nullable;
 
 import java.net.URI;
 import java.util.List;
 import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
 import java.util.concurrent.Executor;
 import java.util.concurrent.ScheduledExecutorService;
-import java.util.concurrent.ScheduledFuture;
 import java.util.concurrent.atomic.AtomicReference;
 
-import static com.facebook.airlift.http.client.HttpUriBuilder.uriBuilderFrom;
-import static com.facebook.airlift.http.client.Request.Builder.prepareGet;
-import static com.facebook.airlift.http.client.Request.Builder.preparePost;
-import static com.facebook.airlift.http.client.StaticBodyGenerator.createStaticBodyGenerator;
-import static com.facebook.presto.execution.TaskInfo.createInitialTask;
 import static com.facebook.presto.execution.buffer.OutputBuffers.BufferType.PARTITIONED;
 import static com.facebook.presto.execution.buffer.OutputBuffers.createInitialEmptyOutputBuffers;
-import static com.facebook.presto.server.RequestHelpers.setContentTypeHeaders;
-import static com.facebook.presto.server.smile.AdaptingJsonResponseHandler.createAdaptingJsonResponseHandler;
-import static com.facebook.presto.spi.StandardErrorCode.NATIVE_EXECUTION_TASK_ERROR;
 import static com.google.common.collect.ImmutableList.toImmutableList;
 import static java.util.Objects.requireNonNull;
-import static java.util.concurrent.TimeUnit.MILLISECONDS;
 
 /**
- * NativeExecutionTask provide the abstraction of executing tasks via c++ worker.
- * The plan is to use it for native execution of Presto on Spark.
- * The plan and splits is provided at creation of NativeExecutionTask, it doesn't
- * support adding more splits during execution.
+ * NativeExecutionTask provide the abstraction of executing tasks via c++ worker. It is used for native execution of Presto on Spark. The plan and splits is provided at creation
+ * of NativeExecutionTask, it doesn't support adding more splits during execution.
+ *
+ * Caller should manage the lifecycle of the task by exposed APIs. The general workflow will look like:
+ * 1. Caller shall start() to start the task. The task finish signal can be told by the returned future.
+ * 2. Caller shall call getTaskInfo() any time to get the current task info.
```

Comment:
QQ for the getTaskInfo(): after the future returned by the `start()` completed, does the `getTaskInfo()` call guarantee to get the final task info (e.g the `status` is `Done`)? Or should the user polling the `getTaskInfo()` until the status in the task info shows it's done?

Reply:
Until caller call stop(), the fetcher will still continuously pull taskinfo from worker and update it in NativeExecutionTask